File: bindings/python/examples/json_decrypt.py
#!/usr/bin/env python3
"""json_decrypt.py — 使用 temari Python 包的完整示例。

演示 **调用方负责网络** 的新模型:
  1. 调用方用 urllib 从 40020 key server 拉取 JSON 响应体
  2. Temari.from_json(json) 解析为模板(库不做任何网络请求)
  3. decrypt / decrypt_par 解密

默认自动拉起仓库 mock 40020,对 10 条测试轨做 from_json 全流程验证,
并逐字节对比已知明文。也可用 `--json` 指定本地 JSON 模板文件(离线)。

运行(需已 `cargo build --release` 构建 libtemari.so):

    cd bindings/python
    python3 examples/json_decrypt.py [--server 127.0.0.1:40020] [--no-mock]
"""
import argparse
import os
import subprocess
import sys
import urllib.parse
import urllib.request

HERE = os.path.dirname(os.path.abspath(__file__))
PKG = os.path.abspath(os.path.join(HERE, ".."))
REPO = os.path.abspath(os.path.join(PKG, "..", ".."))
TESTDATA = os.path.join(REPO, "tests", "testdata")
MOCK = os.path.join(REPO, "tools", "temari-decrypt", "tests", "mock_key_server.py")

sys.path.insert(0, PKG)
from temari import Temari  # noqa: E402

TRACKS = {
    "1720704575": "skd://itunes.apple.com/p683167073/c23",
    "1720704582": "skd://itunes.apple.com/p683167040/c6",
    "1720704586": "skd://itunes.apple.com/p683167043/c6",
    "1720704833": "skd://itunes.apple.com/p683167041/c6",
    "1720704841": "skd://itunes.apple.com/p683166958/c23",
    "1720704847": "skd://itunes.apple.com/p683167009/c6",
    "1720704989": "skd://itunes.apple.com/p683167008/c6",
    "1720704998": "skd://itunes.apple.com/p683167044/c23",
    "1720705006": "skd://itunes.apple.com/p683167074/c23",
    "1720705190": "skd://itunes.apple.com/p683166957/c6",
}


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--server", default="127.0.0.1:40020")
    ap.add_argument("--no-mock", action="store_true", help="不自动拉起 mock 40020")
    ap.add_argument("--json", default=None, help="本地 JSON 模板文件(离线, 跳过网络)")
    args = ap.parse_args()

    mock = None
    if args.json is None and not args.no_mock:
        mock = subprocess.Popen(
            [sys.executable, MOCK, "40020"],
            stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL,
        )
        wait_port(args.server, timeout=10)

    try:
        if args.json:
            run_local(args.json)
        else:
            run_network(args.server)
        print("JSON-DECRYPT: PASS")
    finally:
        if mock:
            mock.terminate()


def fetch_json(server, adam_id, uri):
    """调用方负责网络:从 40020 拉 JSON 响应体(库本身不联网)。"""
    url = f"http://{server}/?adamId={urllib.parse.quote(adam_id)}&uri={urllib.parse.quote(uri)}"
    with urllib.request.urlopen(url, timeout=10) as r:
        return r.read()


def run_network(server):
    for adam, uri in TRACKS.items():
        body = fetch_json(server, adam, uri)          # 网络在这里
        t = Temari.from_json(body)                     # 库只解析
        ct = open(os.path.join(TESTDATA, f"track_{adam}_s1.ct"), "rb").read()
        pt = open(os.path.join(TESTDATA, f"track_{adam}_s1.pt"), "rb").read()
        assert t.decrypt(ct) == pt, f"track {adam}: mismatch"
        t.close()
    print(f"[network] from_json: {len(TRACKS)} tracks PASS (caller-side HTTP)")

    # 并行批量
    t = Temari.from_json(fetch_json(server, "1720704575", TRACKS["1720704575"]))
    ct = open(os.path.join(TESTDATA, "track_1720704575_s1.ct"), "rb").read()
    pt = open(os.path.join(TESTDATA, "track_1720704575_s1.pt"), "rb").read()
    chunks = [ct[i:i + 1024] for i in range(0, len(ct), 1024)]
    plains = t.decrypt_par(chunks)
    assert plains == [t.decrypt(c) for c in chunks]
    t.close()
    print(f"[network] decrypt_par: {len(chunks)} samples PASS")

    # 上下文管理器
    with Temari.from_json(fetch_json(server, "1720704575", TRACKS["1720704575"])) as t:
        assert t.decrypt(ct) == pt
    print("[network] context-manager PASS")

    # 流式:增量提交,按序取回
    t = Temari.from_json(fetch_json(server, "1720704575", TRACKS["1720704575"]))
    expected = [t.decrypt(c) for c in chunks]
    s = t.stream(batch_size=4)
    for c in chunks:
        s.submit(c)
    s.finish()
    got = list(s)
    s.close(); t.close()
    assert got == expected
    print("[network] stream: ordered plaintexts PASS")


def run_local(json_path):
    """离线:读本地 JSON 模板文件(40020 响应体或 gen_testvec 产物)。"""
    with open(json_path, "rb") as f:
        body = f.read()
    t = Temari.from_json(body)
    base = os.path.splitext(os.path.basename(json_path))[0]
    # 若有同名 .ct/.pt 则校验
    ct_p = os.path.join(os.path.dirname(json_path), base + ".ct")
    pt_p = os.path.join(os.path.dirname(json_path), base + ".pt")
    if os.path.exists(ct_p) and os.path.exists(pt_p):
        ct = open(ct_p, "rb").read()
        pt = open(pt_p, "rb").read()
        assert t.decrypt(ct) == pt, "local JSON decrypt mismatch"
        print(f"[local] {json_path}: decrypt==pt PASS")
    t.close()


def wait_port(host_port, timeout=10.0):
    """阻塞直到 host:port 可连接(mock 40020 就绪)。"""
    import socket
    import time
    host, _, port = host_port.rpartition(":")
    port = int(port)
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        try:
            with socket.create_connection((host or "127.0.0.1", port), timeout=1):
                return
        except OSError:
            time.sleep(0.1)
    raise SystemExit(f"server {host_port} did not become ready in {timeout}s")


if __name__ == "__main__":
    main()

File: bindings/go/examples/json_decrypt.go
// json_decrypt.go — temari Go 包完整示例。
//
// 演示 **调用方负责网络** 的新模型:
//  1. 调用方用 net/http 从 40020 key server 拉取 JSON 响应体
//  2. temari.FromJSON(json) 解析为模板(库不做任何网络请求)
//  3. Decrypt / DecryptPar 解密
//
// 默认自动拉起仓库 mock 40020,对 10 条测试轨做 FromJSON 全流程验证,
// 并逐字节对比已知明文。也可用 --json 指定本地 JSON 模板文件(离线)。
//
// 运行(需已 `cargo build --release` 构建 libtemari.so):
//
//	cd <temari>/bindings/go
//	go run ./examples/json_decrypt.go [--server 127.0.0.1:40020] [--no-mock]
package main

import (
	"flag"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"

	"github.com/WorldObservationLog/Temari/bindings/go"
)

var tracks = map[string]string{
	"1720704575": "skd://itunes.apple.com/p683167073/c23",
	"1720704582": "skd://itunes.apple.com/p683167040/c6",
	"1720704586": "skd://itunes.apple.com/p683167043/c6",
	"1720704833": "skd://itunes.apple.com/p683167041/c6",
	"1720704841": "skd://itunes.apple.com/p683166958/c23",
	"1720704847": "skd://itunes.apple.com/p683167009/c6",
	"1720704989": "skd://itunes.apple.com/p683167008/c6",
	"1720704998": "skd://itunes.apple.com/p683167044/c23",
	"1720705006": "skd://itunes.apple.com/p683167074/c23",
	"1720705190": "skd://itunes.apple.com/p683166957/c6",
}

func repoRoot() string {
	_, file, _, _ := runtime.Caller(0)
	// bindings/go/examples/json_decrypt.go -> <repo>/
	return filepath.Clean(filepath.Join(filepath.Dir(file), "..", "..", ".."))
}

func waitPort(hostPort string, timeout time.Duration) error {
	host, port, _ := strings.Cut(hostPort, ":")
	deadline := time.Now().Add(timeout)
	for time.Now().Before(deadline) {
		c, err := net.DialTimeout("tcp", net.JoinHostPort(host, port), time.Second)
		if err == nil {
			c.Close()
			return nil
		}
		time.Sleep(100 * time.Millisecond)
	}
	return fmt.Errorf("server %s not ready in %s", hostPort, timeout)
}

// fetchJSON: 调用方负责网络(库本身不联网)。
func fetchJSON(server, adamID, uri string) ([]byte, error) {
	u := "http://" + server + "/?adamId=" + url.QueryEscape(adamID) + "&uri=" + url.QueryEscape(uri)
	resp, err := http.Get(u)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	return io.ReadAll(resp.Body)
}

func main() {
	server := flag.String("server", "127.0.0.1:40020", "40020 key server")
	noMock := flag.Bool("no-mock", false, "don't spawn the repo mock 40020")
	jsonPath := flag.String("json", "", "local JSON template file (offline)")
	flag.Parse()

	repo := repoRoot()
	testdata := filepath.Join(repo, "tests", "testdata")
	soPath := filepath.Join(repo, "target", "release", "libtemari.so")

	var mock *exec.Cmd
	if *jsonPath == "" && !*noMock {
		mock = exec.Command("python3", filepath.Join(repo, "tools", "temari-decrypt", "tests", "mock_key_server.py"), "40020")
		if err := mock.Start(); err != nil {
			fmt.Println("cannot start mock 40020:", err)
			os.Exit(1)
		}
		defer mock.Process.Kill()
		if err := waitPort(*server, 10*time.Second); err != nil {
			fmt.Println(err)
			os.Exit(1)
		}
	}

	// prefer the cdylib bundled with this module; fall back to the local build
	lib, err := temari.LoadDefault()
	if err != nil {
		lib, err = temari.Load(soPath)
	}
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	if *jsonPath != "" {
		if err := runLocal(lib, *jsonPath); err != nil {
			fmt.Println("FAIL:", err)
			os.Exit(1)
		}
	} else if err := runNetwork(lib, *server, testdata); err != nil {
		fmt.Println("FAIL:", err)
		os.Exit(1)
	}
	fmt.Println("JSON-DECRYPT: PASS")
}

func runNetwork(lib *temari.Library, server, testdata string) error {
	for adam, uri := range tracks {
		body, err := fetchJSON(server, adam, uri) // 网络在这里
		if err != nil {
			return err
		}
		t, err := lib.FromJSON(body) // 库只解析
		if err != nil {
			return err
		}
		ct := readFile(filepath.Join(testdata, "track_"+adam+"_s1.ct"))
		pt := readFile(filepath.Join(testdata, "track_"+adam+"_s1.pt"))
		got, err := t.Decrypt(ct)
		if err != nil {
			return err
		}
		if !equal(got, pt) {
			return fmt.Errorf("track %s: mismatch", adam)
		}
		t.Close()
	}
	fmt.Printf("[network] FromJSON: %d tracks PASS (caller-side HTTP)\n", len(tracks))

	// 并行批量
	body, _ := fetchJSON(server, "1720704575", tracks["1720704575"])
	t, err := lib.FromJSON(body)
	if err != nil {
		return err
	}
	defer t.Close()
	ct := readFile(filepath.Join(testdata, "track_1720704575_s1.ct"))
	pt := readFile(filepath.Join(testdata, "track_1720704575_s1.pt"))
	var chunks [][]byte
	for i := 0; i < len(ct); i += 1024 {
		end := i + 1024
		if end > len(ct) {
			end = len(ct)
		}
		chunks = append(chunks, ct[i:end])
	}
	plains, err := t.DecryptPar(chunks)
	if err != nil {
		return err
	}
	for i, c := range chunks {
		single, _ := t.Decrypt(c)
		if !equal(plains[i], single) {
			return fmt.Errorf("batch sample %d != per-sample", i)
		}
	}
	fmt.Printf("[network] DecryptPar: %d samples PASS\n", len(chunks))

	body2, _ := fetchJSON(server, "1720704575", tracks["1720704575"])
	t2, err := lib.FromJSON(body2)
	if err != nil {
		return err
	}
	got2, _ := t2.Decrypt(ct)
	if !equal(got2, pt) {
		return fmt.Errorf("second handle mismatch")
	}
	t2.Close()
	fmt.Println("[network] multiple handles PASS")

	// 流式:增量提交,按序取回(C() goroutine+channel)
	body3, _ := fetchJSON(server, "1720704575", tracks["1720704575"])
	t3, err := lib.FromJSON(body3)
	if err != nil {
		return err
	}
	s, err := t3.NewStream(4)
	if err != nil {
		return err
	}
	for i := 0; i < len(ct); i += 1024 {
		end := i + 1024
		if end > len(ct) {
			end = len(ct)
		}
		if err := s.Submit(ct[i:end]); err != nil {
			return err
		}
	}
	s.Finish()
	n := 0
	for range s.C() {
		n++
	}
	s.Close()
	t3.Close()
	if n == 0 {
		return fmt.Errorf("stream: no plaintexts")
	}
	fmt.Printf("[network] stream: %d ordered plaintexts PASS\n", n)
	return nil
}

func runLocal(lib *temari.Library, jsonPath string) error {
	body, err := os.ReadFile(jsonPath)
	if err != nil {
		return err
	}
	t, err := lib.FromJSON(body)
	if err != nil {
		return err
	}
	defer t.Close()
	base := strings.TrimSuffix(filepath.Base(jsonPath), filepath.Ext(jsonPath))
	ctP := filepath.Join(filepath.Dir(jsonPath), base+".ct")
	ptP := filepath.Join(filepath.Dir(jsonPath), base+".pt")
	if _, err := os.Stat(ctP); err == nil {
		ct, _ := os.ReadFile(ctP)
		pt, _ := os.ReadFile(ptP)
		got, err := t.Decrypt(ct)
		if err != nil {
			return err
		}
		if !equal(got, pt) {
			return fmt.Errorf("local JSON decrypt mismatch")
		}
		fmt.Printf("[local] %s: decrypt==pt PASS\n", jsonPath)
	}
	return nil
}

func readFile(p string) []byte {
	b, err := os.ReadFile(p)
	if err != nil {
		fmt.Println("read", p, ":", err)
		os.Exit(1)
	}
	return b
}

func equal(a, b []byte) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}
